src/core/lib/iomgr/tcp_posix.c

```diff
@@ -107,7 +112,16 @@ static void tcp_handle_write(grpc_exec_ctx *exec_ctx, void *arg /* grpc_tcp */,
 
 static void tcp_shutdown(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep) {
   grpc_tcp *tcp = (grpc_tcp *)ep;
-  grpc_fd_shutdown(exec_ctx, tcp->em_fd);
+  gpr_mu_lock(&tcp->mu);
```

Comment:
I believe it should just work. It's supposed to work -- if your notification observes the inserted edge on shutdown, you should get an error and follow the !success path of tcp_handle_read.

This is no different than if you directly call shutdown concurrently/after a read or write.

The one requirement is that you don't do it twice -- our implementation only injects a single extra shutdown edge, so the second one can get stuck if it doesn't observe a socket error.